proposals/spi-proposal.md

```diff
@@ -112,168 +105,150 @@ MSBFIRST = 1
+the CS pin using `DIGITAL_MESSAGE` commands, set `CS_PIN_CONTROL` to `disable`.
 
+### SPI_TRANSFER
 
-### SPI_END_TRANSACTION
+Full-duplex write/read transfer. This is the normal SPI transfer mode, a word
+must be written for every word read. Reply is sent via `SPI_REPLY`.
 
-Signals the end of a transaction. Must be sent before starting communication
-with another SPI device.
+Since transport (Serial, Ethernet, Wi-Fi, etc) buffers tend to be small on
+microcontroller platforms, it may be necessary to send several `SPI_TRANSFER`
+commmands to complete a single SPI transaction. Use the `csPinChange` parameter
+to ensure the CS pin is not deselected in between `SPI_TRANSFER` commands until
+the transaction is complete.
 
-```
-0:  START_SYSEX
-1:  SPI_DATA              (0x68)
-2:  SPI_END_TRANSACTION   (0x02)
-3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
-4:  END_SYSEX
-```
-
-### SPI_TRANSFER
+`requestId` is used in the request messages `SPI_TRANSFER`, `SPI_WRITE` and
+`SPI_READ` and in the reply message `SPI_REPLY`. Its purpose is to ensure that
+the SPI_REPLY message matches the request. For each request message, increment
+a single 7-bit requestId value, rolling it over to 0 when > 127.
 
-Full-duplex write/read transfer. This is the normal SPI transfer mode, a word must be
-written for every word read. Reply is sent via `SPI_REPLY`.
+`csPinChange` is used to control the csPin at the end of the transfer. By
+default the csPin will be deselected (the csPin will be toggled). However, to
+prevent deselection to enable back-to-back transfers for example, set
+`csPinChange` to 0 and the pin state won't be affected.
 
-The user can continue sending as many transfer messages as necessary, even
-alternating between `SPI_TRANSFER`, `SPI_WRITE` and `SPI_READ` within a single SPI
-transaction.
+If `CS_PIN_CONTROL` is enabled, then the CS pin active state will be set when
+the `SPI_TRANSFER` command is received. It will only be deselected (toggled) at
+the end of the transfer if `csPinChange` is set to 1.
 
 ```
 0:  START_SYSEX
 1:  SPI_DATA              (0x68)
-2:  SPI_TRANSFER          (0x03)
-3:  channel               (HW supports multiple SPI ports. range = 0-3, default = 0)
-4:  csPinControl          (see details in csPinControl section below)
-                            0x00: default behavior (as specified by csPinOptions)
-                            0x01: CS_DISABLE
-                            0x02: CS_START_ONLY
-                            0x03: CS_END_ONLY
-5.  numWords              (number of words to transfer)
-6:  data 0                (bits 0-6)
-7:  data 0                (bits 7-14 if word size if word size > 7 && < 15)
-8:  data 0                (if word size > 14)
+2:  SPI_TRANSFER          (0x02)
+3:  deviceId | channel    (bits 3-6: deviceId, bits 0-2: channel)
+4:  requestId             (0-127) // increment for each call
+5:  csPinChange           (0 = don't deselect csPin
```

Comment:
`csPinChange` could be ignored for Linux-based boards, but it's necessary for Arduino-compatible and perhaps other non-Linux-based platforms

Reply:
I would also love to stuff this bit somewhere rather than allocating an entire byte for it. I could stuff it in requestId, but that would limit the requestId range to 0-63. I could stuff it in byte 3 but that would require limiting the deviceId range to 8 values or the channel range to 4 values (keeping device Id as 16).
However, by keeping it as byte 5 (as above) the other bits in byte 5 can be reserved in case additional functionality is needed in the future.

Reply:
To confirm confirm that we're on the same page here, `csPinChange` will be ignored for Linux-based boards if character special device files as described [here](https://www.kernel.org/doc/Documentation/spi/spidev) are used to access SPI devices. However, if firmata.js and johnny-five (including io-plugins) exposed separate spiTransfer and spiMessage methods as described [here](https://github.com/firmata/protocol/pull/105#issuecomment-399725889) where spiTransfer is a single transfer and spiMessage is for multi-transfer then `csPinChange` could be used as intended for multi-transfers. Is this correct?

Reply:
This is tricky. There are 2 different concepts of csPinChange, one that's part of Firmata and has to be passed over the wire, and another than's specific to Linux. They do the same thing, but one has to go over a wire and one doesn't.

Reply:
On Linux, I guess csPinChange would be a no-op for spiTransfer. That may be difficult to communicate as it may not align with a user's expectations. I'm not sure of a better way to handle it since I think having that functionality in spiTransfer simplifies things for Arduino or other transport-based Firmata implementations.